capi/cbindgen.toml

```diff
@@ -38,6 +38,9 @@ exclude = ["Addr"]
 "Normalizer" = "struct blaze_normalizer"
 "Symbolizer" = "struct blaze_symbolizer"
 
+[layout]
+packed = "__attribute__((packed))"
```

Comment:
oh, I wouldn't do packed... (I've been told) it forces compiler to generate less optimal code when accessing members of the struct (due to assumption that it can be at any unaligned address).

Let's drop packed, if we want to make sure we don't have any padding, we can use explicit "do-not-use-or-reserved" fields to make sure structs are zero initialized properly

Reply:
I started off with non-packed and it's just a huge *()!#&*($!@#*()(#!*!)(:
- it requires this `_last_field` dance
- it requires `offset_of!` macro, which Rust doesn't currently have built-in (requires additional dependency; though it's slated to stabilize with 1.77, I think)
- I don't think it works reliably in Rust. We can zero initialize a struct from a zeroed byte buffer, but the moment we copy it (which happens way more often than in C, mind you) the compiler may fill padding/trailing bytes with arbitrary data, breaking the logic; I would not be surprised if the same is true for C, by the way, it's just that there copying a struct using built-in copy syntax is rarely ever done, and presumably never for the types that `libbpf` uses it for
  - yes, these types are meant for C code, but still, we need them in Rust for testing purposes

These aren't performance critical struct by any means. At least on x86, I'd think the generated code is likely the same, only that loads could potentially be penalized if unaligned. But again, the types that are packed are option types that are not in a performance critical path.

Reply:
I suppose making padding bytes explicit is an option. But it's still more error prone, I'd say, because it's easy to forget/get wrong and not easily automatically verifiable -- whereas with packed we know there is nothing else.
Let me know how strongly you feel.

Reply:
I just feel uneasy about this packed thing, that's all. I'm not saying it will ever cause problems, but my gut feel says it might, which is why I'm pushing back a bit. If you feel confident about this approach, go for it.

last_field isn't a big deal, you don't change that often and neither there are a lot of structs with size field, for offset_of(), until it stabilizes we can use extra crate of create our own version of offset_of() (not sure how involved it is, but there seems to be https://github.com/Gilnaa/memoffset/ available). Explicit padding at the end is indeed error-prone, especially taking into account 32-bit vs 64-bit architectures, but it's all doable.

Again, this is just a gut feel and I don't think I've seen many libraries that use packed structs on their "main" types.

re: performance, you might not penalize x86-64 much, but you probably can on arm64 (but again, I don't have much knowledge about arm64 specifics). Anyways, I'm not going to fight for this, but I'd probably not opt into packed struct myself.

Reply:
I removed `packed` and made trailing padding explicit now.